awx/main/scheduler/task_manager.py

```diff
@@ -598,6 +555,8 @@ def process_tasks(self, all_sorted_tasks):
         self.process_running_tasks(running_tasks)
 
         pending_tasks = [t for t in all_sorted_tasks if t.status == 'pending']
+        dependencies = self.generate_dependencies(pending_tasks)
+        self.process_pending_tasks(dependencies)
         self.process_pending_tasks(pending_tasks)
```

Comment:
@AlanCoding 

removed recursion. generate_dependencies() now takes in a list. Maybe I should append pending_tasks to dependencies and call process_pending_tasks() once.

Reply:
I actually like this a lot better. Someone reading the code much more quickly gets a high-level understanding of what's going on. The need for processing pending dependencies before pending tasks is really clear.

The only hazard I'm seeing right now is the concern I mentioned in the comment

https://github.com/ansible/awx/pull/5672/files#r367409144

I'm worried that a project update will exist in both `dependencies` and `pending_tasks` here, as 2 separate instances in-memory (but same entry in DB). If the first run of `process_pending_tasks` switches it from pending to running, the 2nd run will not know about this, and have the stale status of "pending" on its version of that project update. We would like to avoid those situations at all costs. You could prune `pending_tasks` of any entries inside the `dependencies` list, or reorganize so that you separate the newly-created dependencies from the just-linked dependencies.